crates/fuzzing/src/oracles/v8.rs

```diff
@@ -0,0 +1,289 @@
+            .get_memory(&mut wasmtime_store, mem)
+            .unwrap();
+
+        let name = v8::String::new(&mut scope, "MEMORY_NAME").unwrap();
+        let func_name = v8::String::new(&mut scope, mem).unwrap();
+        global.set(&mut scope, name.into(), func_name.into());
+        let v8 = eval(
+            &mut scope,
+            &format!("WASM_INSTANCE.exports[MEMORY_NAME].buffer"),
+        )
+        .unwrap();
+        let v8 = v8::Local::<'_, v8::ArrayBuffer>::try_from(v8).unwrap();
+        let v8_data = v8.get_backing_store();
+        let wasmtime_data = wasmtime.data(&wasmtime_store);
+        assert_eq!(wasmtime_data.len(), v8_data.len());
+        for i in 0..v8_data.len() {
+            if wasmtime_data[i] != v8_data[i].get() {
+                panic!("memories differ");
+            }
+        }
+    }
+
+    Some(())
+}
+
+/// Manufactures a new V8 Isolate to run within.
+fn isolate() -> v8::OwnedIsolate {
+    static INIT: Once = Once::new();
+
+    INIT.call_once(|| {
+        let platform = v8::new_default_platform(0, false).make_shared();
+        v8::V8::initialize_platform(platform);
+        v8::V8::initialize();
+    });
+
+    v8::Isolate::new(Default::default())
+}
+
+/// Evaluates the JS `code` within `scope`, returning either the result of the
+/// computation or the stringified exception if one happened.
+fn eval<'s>(
+    scope: &mut v8::HandleScope<'s>,
+    code: &str,
+) -> Result<v8::Local<'s, v8::Value>, String> {
+    let mut tc = v8::TryCatch::new(scope);
+    let mut scope = v8::EscapableHandleScope::new(&mut tc);
+    let source = v8::String::new(&mut scope, code).unwrap();
+    let script = v8::Script::compile(&mut scope, source, None).unwrap();
+    match script.run(&mut scope) {
+        Some(val) => Ok(scope.escape(val)),
+        None => {
+            drop(scope);
+            assert!(tc.has_caught());
+            Err(tc
+                .message()
+                .unwrap()
+                .get(&mut tc)
+                .to_rust_string_lossy(&mut tc))
+        }
+    }
+}
+
+/// Asserts that the wasmtime value `a` matches the v8 value `b`.
+///
+/// For NaN values simply just asserts that they're both NaN.
+fn assert_val_match(a: &Val, b: &v8::Local<'_, v8::Value>, scope: &mut v8::HandleScope<'_>) {
+    match *a {
+        Val::I32(wasmtime) => {
+            assert_eq!(i64::from(wasmtime), b.to_int32(scope).unwrap().value());
+        }
+        Val::I64(wasmtime) => {
+            assert_eq!((wasmtime, true), b.to_big_int(scope).unwrap().i64_value());
+        }
+        Val::F32(wasmtime) => {
+            same_float(
+                f64::from(f32::from_bits(wasmtime)),
+                b.to_number(scope).unwrap().value(),
+            );
+        }
+        Val::F64(wasmtime) => {
+            same_float(
+                f64::from_bits(wasmtime),
+                b.to_number(scope).unwrap().value(),
+            );
+        }
+        _ => panic!("unsupported match {:?}", a),
```

Comment:
Shouldn't we be comparing `v128` here and above?

Reply:
Node doesn't, at this time, support v128 values, so if you have a function that returns `v128`, for example, you get:

```
TypeError: type incompatibility when transforming from/to JS
```

Above this though there's a check with:

```
    // not supported yet in V8
    if ty.params().chain(ty.results()).any(|t| t == ValType::V128) {
        log::trace!("exported function uses v128, skipping");
        return None;
    }
```

to avoid hitting this case by just skipping that fuzzing candidate.

Reply:
er I should say that it doesn't support v128-in-JS, it only supports v128-in-wasm, so because we can only interact to wasm through JS that's where the limitation comes from

Reply:
Ah, I see.